File: src/main/java/org/jdc/template/ui/activity/IndividualEditActivity.java
package org.jdc.template.ui.activity;

import android.app.DatePickerDialog;
import android.app.TimePickerDialog;
import android.os.Bundle;
import android.support.design.widget.TextInputEditText;
import android.support.design.widget.TextInputLayout;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.Toolbar;
import android.text.format.DateUtils;
import android.view.Menu;
import android.view.MenuItem;

import com.google.android.gms.analytics.HitBuilders;

import org.apache.commons.lang3.StringUtils;
import org.dbtools.android.domain.date.DBToolsThreeTenFormatter;
import org.jdc.template.Analytics;
import org.jdc.template.R;
import org.jdc.template.inject.Injector;
import org.jdc.template.model.database.main.individual.Individual;
import org.jdc.template.model.database.main.individual.IndividualManager;
import org.threeten.bp.LocalDate;
import org.threeten.bp.LocalTime;
import org.threeten.bp.ZoneId;

import javax.annotation.Nonnull;
import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import pocketknife.BindExtra;
import pocketknife.PocketKnife;

public class IndividualEditActivity extends BaseActivity {

    public static final String EXTRA_ID = "INDIVIDUAL_ID";

    @BindExtra(EXTRA_ID)
    long individualId;

    @BindView(R.id.ab_toolbar)
    Toolbar toolbar;
    @BindView(R.id.first_name_layout)
    TextInputLayout firstNameLayout;
    @BindView(R.id.alarm_time_layout)
    TextInputLayout alarmTimeLayout;
    @BindView(R.id.first_name)
    TextInputEditText firstNameEditText;
    @BindView(R.id.last_name)
    TextInputEditText lastNameEditText;
    @BindView(R.id.phone)
    TextInputEditText phoneEditText;
    @BindView(R.id.email)
    TextInputEditText emailEditText;
    @BindView(R.id.birth_date)
    TextInputEditText birthDateEditText;
    @BindView(R.id.alarm_time)
    TextInputEditText alarmTimeEditText;

    @Inject
    IndividualManager individualManager;
    @Inject
    Analytics analytics;

    @Nonnull
    private Individual individual = new Individual();

    private DatePickerDialog birthDatePickerDialog;
    private TimePickerDialog alarmTimePickerDialog;

    public IndividualEditActivity() {
        Injector.get().inject(this);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_individual_edit);
        ButterKnife.bind(this);
        PocketKnife.bindExtras(this);

        setupActionBar();

        showIndividual();
    }

    private void setupActionBar() {
        setSupportActionBar(toolbar);
        enableActionBarBackArrow();
        ActionBar actionBar = getSupportActionBar();

        if (actionBar != null) {
            actionBar.setTitle(R.string.individual);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.individual_edit_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item == null) {
            return false;
        }

        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                return true;
            case R.id.menu_item_save:
                saveIndividual();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }


    @OnClick(R.id.birth_date)
    public void onBirthdayClick() {
        if (birthDatePickerDialog == null) {

            LocalDate date = individual.getBirthDate() != null ? individual.getBirthDate() : LocalDate.now();
            birthDatePickerDialog = new DatePickerDialog(this, (view, year, monthOfYear, dayOfMonth) -> {
                individual.setBirthDate(LocalDate.of(year, monthOfYear + 1, dayOfMonth)); // + 1 because cord Java Date is 0 based
                showBirthDate();
            }, date.getYear(), date.getMonthValue() - 1, date.getDayOfMonth()); // - 1 because cord Java Date is 0 based
        }

        birthDatePickerDialog.show();
    }

    @OnClick(R.id.alarm_time)
    public void onAlarmClick() {
        if (alarmTimePickerDialog == null) {

            LocalTime time = individual.getAlarmTime() != null ? individual.getAlarmTime() : LocalTime.now();
            alarmTimePickerDialog = new TimePickerDialog(this, (view, hourOfDay, minute) -> {
                individual.setAlarmTime(LocalTime.of(hourOfDay, minute));
                showAlarmTime();
            }, time.getHour(), time.getMinute(), false);
        }

        alarmTimePickerDialog.show();
    }

    private void showIndividual() {
        if (individualId <= 0) {
            return;
        }

        Individual foundIndividual = individualManager.findByRowId(individualId);
        if (foundIndividual != null) {
            individual = foundIndividual;
            analytics.send(new HitBuilders.EventBuilder()
                    .setCategory(Analytics.CATEGORY_INDIVIDUAL)
                    .setAction(Analytics.ACTION_EDIT)
                    .build());

            firstNameEditText.setText(individual.getFirstName());
            lastNameEditText.setText(individual.getLastName());
            emailEditText.setText(individual.getEmail());
            phoneEditText.setText(individual.getPhone());

            showBirthDate();
            showAlarmTime();
        }
    }

    private void showBirthDate() {
        if (individual.getBirthDate() == null) {
            return;
        }

        LocalDate date = individual.getBirthDate();
        long millis = DBToolsThreeTenFormatter.localDateTimeToLong(date.atStartOfDay(ZoneId.systemDefault()).toLocalDateTime());
        birthDateEditText.setText(DateUtils.formatDateTime(this, millis, DateUtils.FORMAT_SHOW_DATE | DateUtils.FORMAT_SHOW_YEAR));
    }

    private void showAlarmTime() {
        if (individual.getAlarmTime() == null) {
            return;
        }

        LocalTime time = individual.getAlarmTime();
        long millis = DBToolsThreeTenFormatter.localDateTimeToLong(time.atDate(LocalDate.now()));
        alarmTimeEditText.setText(DateUtils.formatDateTime(this, millis, DateUtils.FORMAT_SHOW_TIME));
    }

    private void saveIndividual() {
        if (StringUtils.isBlank(firstNameEditText.getText())) {
            firstNameLayout.setError(getString(R.string.required));
            return;
        }

        if (individual.getAlarmTime() == null) {
            alarmTimeLayout.setError(getString(R.string.required));
            return;
        }

        individual.setFirstName(firstNameEditText.getText().toString());
        individual.setLastName(lastNameEditText.getText().toString());
        individual.setPhone(phoneEditText.getText().toString());
        individual.setEmail(emailEditText.getText().toString());

        individualManager.save(individual);

        finish();
    }
}

File: src/main/kotlin/org/jdc/template/ui/activity/IndividualEditActivity.kt
package org.jdc.template.ui.activity

import android.app.Activity
import android.app.DatePickerDialog
import android.app.TimePickerDialog
import android.os.Bundle
import android.text.format.DateUtils
import android.view.Menu
import android.view.MenuItem
import com.google.android.gms.analytics.HitBuilders
import kotlinx.android.synthetic.main.individual_edit.*
import kotlinx.android.synthetic.main.toolbar_actionbar.*
import org.dbtools.android.domain.date.DBToolsThreeTenFormatter
import org.jdc.template.Analytics
import org.jdc.template.R
import org.jdc.template.inject.Injector
import org.jdc.template.model.database.main.individual.Individual
import org.jdc.template.model.database.main.individual.IndividualManager
import org.threeten.bp.LocalDate
import org.threeten.bp.LocalTime
import org.threeten.bp.ZoneId
import pocketknife.BindExtra
import pocketknife.PocketKnife
import javax.inject.Inject

class IndividualEditActivity : BaseActivity() {

    @BindExtra(EXTRA_ID)
    var individualId = 0L

    @Inject
    lateinit var individualManager: IndividualManager
    @Inject
    lateinit var analytics: Analytics

    private var individual = Individual()

    private var birthDatePickerDialog: DatePickerDialog? = null
    private var alarmTimePickerDialog: TimePickerDialog? = null

    init {
        Injector.get().inject(this)
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_individual_edit)
        PocketKnife.bindExtras(this)

        setSupportActionBar(ab_toolbar)
        enableActionBarBackArrow(true)

        setDefaultKeyMode(Activity.DEFAULT_KEYS_SEARCH_LOCAL)

        setupActionBar()

        alarmTimeEditText.setOnClickListener {
            onAlarmClick()
        }

        birthDateEditText.setOnClickListener {
            onBirthdayClick()
        }

        showIndividual()

    }

    private fun setupActionBar() {
        setSupportActionBar(ab_toolbar)
        val actionBar = supportActionBar

        actionBar?.setTitle(R.string.individual)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.individual_edit_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                finish()
                return true
            }
            R.id.menu_item_save -> {
                saveIndividual()
                return true
            }

            else -> return super.onOptionsItemSelected(item)
        }
    }

    fun onBirthdayClick() {
        if (birthDatePickerDialog == null) {

            val date = if (individual.birthDate != null) individual.birthDate else LocalDate.now()
            if (date != null) {
                birthDatePickerDialog = DatePickerDialog(this, DatePickerDialog.OnDateSetListener { view, year, monthOfYear, dayOfMonth ->
                    individual.birthDate = LocalDate.of(year, monthOfYear + 1, dayOfMonth) // + 1 because cord Java Date is 0 based
                    showBirthDate()
                }, date.year, date.monthValue - 1, date.dayOfMonth) // - 1 because cord Java Date is 0 based
            }
        }

        birthDatePickerDialog!!.show()
    }

    fun onAlarmClick() {
        if (alarmTimePickerDialog == null) {
            val time = if (individual.alarmTime != null) individual.alarmTime else LocalTime.now()
            alarmTimePickerDialog = TimePickerDialog(this, TimePickerDialog.OnTimeSetListener { view, hourOfDay, minute ->
                individual.alarmTime = LocalTime.of(hourOfDay, minute)
                showAlarmTime()
            }, time.hour, time.minute, false)
        }

        alarmTimePickerDialog!!.show()
    }

    private fun showIndividual() {
        if (individualId <= 0) {
            return
        }

        val foundIndividual = individualManager.findByRowId(individualId)
        if (foundIndividual != null) {
            individual = foundIndividual
            analytics.send(HitBuilders.EventBuilder().setCategory(Analytics.CATEGORY_INDIVIDUAL).setAction(Analytics.ACTION_EDIT).build())

            firstNameEditText.setText(individual.firstName)
            lastNameEditText.setText(individual.lastName)
            emailEditText.setText(individual.email)
            phoneEditText.setText(individual.phone)

            showBirthDate()
            showAlarmTime()
        }
    }

    private fun showBirthDate() {
        if (individual.birthDate == null) {
            return
        }

        val date = individual.birthDate
        if (date != null) {
            val millis = DBToolsThreeTenFormatter.localDateTimeToLong(date.atStartOfDay(ZoneId.systemDefault()).toLocalDateTime())
            birthDateEditText.setText(DateUtils.formatDateTime(this, millis!!, DateUtils.FORMAT_SHOW_DATE or DateUtils.FORMAT_SHOW_YEAR))
        }
    }

    private fun showAlarmTime() {
        if (individual.alarmTime == null) {
            return
        }

        val time = individual.alarmTime
        val millis = DBToolsThreeTenFormatter.localDateTimeToLong(time.atDate(LocalDate.now()))
        alarmTimeEditText.setText(DateUtils.formatDateTime(this, millis!!, DateUtils.FORMAT_SHOW_TIME))
    }

    private fun saveIndividual() {
        if (firstNameEditText.text.isBlank()) {
            firstNameLayout.error = getString(R.string.required)
            return
        }

        if (individual.alarmTime == null) {
            alarmTimeLayout.error = getString(R.string.required)
            return
        }

        individual.firstName = firstNameEditText.text.toString()
        individual.lastName = lastNameEditText.text.toString()
        individual.phone = phoneEditText.text.toString()
        individual.email = emailEditText.text.toString()

        individualManager.save(individual)

        finish()
    }

    companion object {

        const val EXTRA_ID = "INDIVIDUAL_ID"
    }
}